File: solutions/haskell/bob/1/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where
import qualified Data.Text as T
import Data.Text (Text)
import Data.Char (isUpper, isSpace, isAlpha)


responseFor :: Text -> Text
responseFor question
  | noSpace == "" = "Fine. Be that way!"
  | isYelling && not isQuestion = "Whoa, chill out!"
  | isYelling = "Calm down, I know what I'm doing!"
  | isQuestion = "Sure."
  | otherwise = "Whatever."
  where noSpace = T.filter (\c -> not (isSpace c)) question
        alphabets = T.filter isAlpha noSpace
        isYelling =  not (T.null alphabets) && T.all isUpper alphabets
        isQuestion = T.isSuffixOf "?" noSpace

File: solutions/typescript/bob/1/bob.ts
class Msg {
  isQuestion;
  isYell;
  isSilence;
  constructor(message:string) {
    this.isQuestion = message.trim().endsWith("?");
    this.isYell = message.toUpperCase() === message && message.search(/[a-zA-Z]/) !== -1;
    this.isSilence = message.trim().length === 0;
  }
}

export function hey(message: string): string {
  const msg = new Msg(message);
  if (msg.isSilence) return `Fine. Be that way!`;
  if (msg.isQuestion && !msg.isYell) return `Sure.`
  if (!msg.isQuestion && msg.isYell) return `Whoa, chill out!`
  if (msg.isQuestion && msg.isYell) return `Calm down, I know what I'm doing!`;
  return `Whatever.`
}
